orderer/common/cluster/clusterservice.go

```diff
@@ -0,0 +1,210 @@
+type ClusterStepStream interface {
+	Send(response *orderer.ClusterNodeServiceStepResponse) error
+	Recv() (*orderer.ClusterNodeServiceStepRequest, error)
+	grpc.ServerStream
+}
+
+// ClusterService defines the raft Service
+type ClusterService struct {
+	StreamCountReporter              *StreamCountReporter
+	Dispatcher                       Dispatcher
+	Logger                           *flogging.FabricLogger
+	StepLogger                       *flogging.FabricLogger
+	MinimumExpirationWarningInterval time.Duration
+	CertExpWarningThreshold          time.Duration
+	NodesCert                        map[uint64][]byte
+}
+
+// Step passes an implementation-specific message to another cluster member.
+func (s *ClusterService) Step(stream orderer.ClusterNodeService_StepServer) error {
+	s.StreamCountReporter.Increment()
+	defer s.StreamCountReporter.Decrement()
+
+	addr := util.ExtractRemoteAddress(stream.Context())
+	commonName := commonNameFromContext(stream.Context())
+	exp := s.initializeExpirationCheck(stream, addr, commonName)
+	s.Logger.Debugf("Connection from %s(%s)", commonName, addr)
+	defer s.Logger.Debugf("Closing connection from %s(%s)", commonName, addr)
+
+	//expect auth message is the first msg on the stream
+	request, err := stream.Recv()
+	if err == io.EOF {
+		s.Logger.Debugf("%s(%s) disconnected well before establishing the stream", commonName, addr)
+		return nil
+	}
+	if err != nil {
+		s.Logger.Warningf("Stream read from %s failed: %v", addr, err)
+		return err
+	}
+	channel, err := s.VerifyAuthRequest(stream, request)
+	if err != nil {
+		s.Logger.Warnf("gRPC Stream authentication failed with error: %v", err)
+		return err
+	}
+
+	for {
+		err := s.handleMessage(stream, addr, exp, channel)
+		if err == io.EOF {
+			s.Logger.Debugf("%s(%s) disconnected", commonName, addr)
+			return nil
+		}
+		if err != nil {
+			return err
+		}
+		// Else, no error occurred, so we continue to the next iteration
+	}
+}
+
+func (s *ClusterService) VerifyAuthRequest(stream orderer.ClusterNodeService_StepServer, request *orderer.ClusterNodeServiceStepRequest) (string, error) {
+	authReq := request.GetNodeAuthrequest()
+	if authReq == nil {
+		return "", status.Error(codes.Unauthenticated, "Authentication failed. Message is out of order")
+	} else {
+		bindingFieldsHash := util.ComputeSHA256(util.ConcatenateBytes(
+			[]byte(strconv.FormatUint(uint64(authReq.Version), 10)),
+			[]byte(authReq.Timestamp.String()),
+			[]byte(strconv.FormatUint(authReq.FromId, 10)),
+			[]byte(strconv.FormatUint(authReq.ToId, 10)),
+			[]byte(authReq.Channel),
+		))
+
+		tlsBinding, err := GetTLSSessionBinding(stream, bindingFieldsHash)
+		if err != nil {
+			status.Errorf(codes.Unauthenticated, "Session binding error %v", err)
+			return authReq.Channel, err
+		}
+
+		if bytes.Equal(tlsBinding, authReq.SessionBinding) {
+			return authReq.Channel, status.Error(codes.Unauthenticated, "Session binding mismatch")
+		}
+
+		cert := s.NodesCert[authReq.FromId]
```

Comment:
It should be not only the cert but a serialized identtiy, so that later we could plug in the MSP. 

serializedIdenttiy is a protobuf of msp package which is a tuple of MSP ID and a PEM encoded x509 certificate.

Reply:
updated with github.com/hyperledger/fabric-protos-go/msp.SerializedIdentity